src/main/java/frc/robot/subsystems/Indexer.java

```diff
@@ -0,0 +1,220 @@
+    private static final double pierreIndexSpeed = 0.85;
+
+    public Indexer(final IndexMap map) {
+        super();
+        singulator = map.singulator();
+        frontIntakeIR = map.frontIntakeIR();
+        bottomPierreIR = map.bottomPierreIR();
+        topPierreIR = map.topPierreIR();
+        backIntakeIR = map.backIntakeIR();
+        pierreMotor = map.pierreMotor();
+
+    }
+
+    public SequentialCommandGroup intakeToPierre() {
+        return new SequentialCommandGroup(pierrePossesion(), runToClearBottomSensor());
+    }
+
+    public SequentialCommandGroup shootOneBall() {
+        return new SequentialCommandGroup(loadBallToTop(), unLoadBall());
+    }
+
+    public ParallelDeadlineGroup topSensorTriggered() {
+        final ParallelDeadlineGroup topIRStop = new ParallelDeadlineGroup(
+                new SequentialCommandGroup(pierrePossesion(), runToClearBottomSensor()));
+        topIRStop.setDeadline(stopWhenBallsAtTop());
+        return topIRStop;
+    }
+    // This will stop all commands when the top IR sensor on Pierre is triggered
+
+    public CommandBase indexMotor(final double motorSpeed) {
+        return new StartEndCommand(() -> {
+            singulator.set(motorSpeed);
+        }, () -> {
+            singulator.set(0);
+        }, this);
+    }
+
+    public CommandBase quicklyPush() {
+        return indexMotor(singulatorMotorSpeed);
+    }
+
+    public CommandBase reversePush() {
+        return indexMotor(-singulatorMotorSpeed);
+    }
+
+    public CommandBase quicklyOutput() {
+        return indexMotor(pierreIndexSpeed);
+    }
+
+    public CommandBase reverseOutput() {
+        return indexMotor(-pierreIndexSpeed);
+    }
+
+    public CommandBase singulatorPossesion() {
+        return new FunctionalCommand(() -> {
+
+        }, () -> {
+
+        }, (interrupted) -> {
+
+            singulator.set(0);
+
+        }, () -> {
+
+            return frontIntakeIR.getAsBoolean() || topPierreIR.getAsBoolean();
+
+        }, this);
+
+    }
+    // This command will make sure that the singulator has possesion of the ball
+
+    public CommandBase pierrePossesion() {
+        return new FunctionalCommand(() -> {
+
+            if (frontIntakeIR.getAsBoolean()) {
+
+                singulator.set(singulatorMotorSpeed);
+            }
+            if (backIntakeIR.getAsBoolean()) {
+                pierreMotor.set(pierreIndexSpeed);
+                singulator.set(singulatorMotorSpeed);
+            }
+
+        }, () -> {
+
+            if (frontIntakeIR.getAsBoolean()) {
+
+                singulator.set(singulatorMotorSpeed);
+            }
+            if (backIntakeIR.getAsBoolean()) {
+                pierreMotor.set(pierreIndexSpeed);
+                singulator.set(singulatorMotorSpeed);
+            }
+        }, (interrupted) -> {
+
+            pierreMotor.set(0);
+            singulator.set(0);
+            SmartDashboard.putNumber("Ball Count", ballCounting);
```

Comment:
Where is `ballCounting` set?

If this needs to be set all the time, override the `periodic` function and do it there.

Reply:
We will deal with this in the future with OBLog